File: tinyml/scripts/feature_extractor.py
"""
Módulo compartido: extractor de features del pipeline ML.

Este archivo es importado tanto por los scripts de entrenamiento
como por el backend durante la inferencia.

Uso:
    from feature_extractor import extract_features_from_window

    features_dict = extract_features_from_window(ax_list, ay_list, az_list)
    feature_vector = list(features_dict.values())  # para el modelo
"""

import numpy as np


AXIS_NAMES = ["ax", "ay", "az"]


def extract_features_from_window(
    ax: list | np.ndarray,
    ay: list | np.ndarray,
    az: list | np.ndarray,
) -> dict:
    """
    Calcula las features estadísticas de una ventana de aceleración.

    Args:
        ax: Lista o array de muestras de aceleración en X
        ay: Lista o array de muestras de aceleración en Y
        az: Lista o array de muestras de aceleración en Z

    Returns:
        Diccionario ordenado {nombre_feature: valor_float}
        Listo para construir el vector de entrada al modelo.

    Features (25 en total):
        Por eje (×3): mean, std, min, max, range, rms, energy  → 21
        De magnitud : magnitude_mean, magnitude_std, magnitude_max → 3
        Global      : zcr_ax → 1
    """
    ax_a = np.asarray(ax, dtype=np.float64)
    ay_a = np.asarray(ay, dtype=np.float64)
    az_a = np.asarray(az, dtype=np.float64)

    features = {}

    for name, vals in zip(AXIS_NAMES, [ax_a, ay_a, az_a]):
        features[f"{name}_mean"] = float(np.mean(vals))
        features[f"{name}_std"] = float(np.std(vals))
        features[f"{name}_min"] = float(np.min(vals))
        features[f"{name}_max"] = float(np.max(vals))
        features[f"{name}_range"] = float(np.max(vals) - np.min(vals))
        features[f"{name}_rms"] = float(np.sqrt(np.mean(vals**2)))
        features[f"{name}_energy"] = float(np.sum(vals**2) / len(vals))

    # Magnitud del vector 3D
    magnitude = np.sqrt(ax_a**2 + ay_a**2 + az_a**2)
    features["magnitude_mean"] = float(np.mean(magnitude))
    features["magnitude_std"] = float(np.std(magnitude))
    features["magnitude_max"] = float(np.max(magnitude))

    # Tasa de cruces por cero en ax (detecta oscilaciones periódicas)
    centered = ax_a - ax_a.mean()
    crossings = np.where(np.diff(np.sign(centered)))[0]
    features["zcr_ax"] = float(len(crossings) / len(ax_a))

    return features


def feature_vector(ax, ay, az) -> list[float]:
    """
    Retorna la lista de valores (sin nombres) para pasar directo al modelo.
    El orden es el mismo que el usado durante el entrenamiento.
    """
    return list(extract_features_from_window(ax, ay, az).values())


def feature_names() -> list[str]:
    """Retorna los nombres de todas las features en el mismo orden."""
    dummy = np.zeros(10)
    return list(extract_features_from_window(dummy, dummy, dummy).keys())

File: tinyml/components/feature_extractor/feature_extractor.c
/**
 * @file feature_extractor.c
 * @brief Implementación C de las 25 features estadísticas del pipeline ML.
 *
 * Mantiene el mismo orden y cálculo que feature_extractor.py del proyecto ml/.
 */

#include "feature_extractor.h"

#include <math.h>
#include "esp_err.h"

/* ── Helpers estadísticos ────────────────────────────────────── */

static float _mean(const float *v, int n)
{
    float s = 0.0f;
    for (int i = 0; i < n; i++) s += v[i];
    return s / (float)n;
}

static float _std(const float *v, int n, float mean)
{
    float s = 0.0f;
    for (int i = 0; i < n; i++) {
        float d = v[i] - mean;
        s += d * d;
    }
    return sqrtf(s / (float)n);
}

static float _min(const float *v, int n)
{
    float m = v[0];
    for (int i = 1; i < n; i++) if (v[i] < m) m = v[i];
    return m;
}

static float _max(const float *v, int n)
{
    float m = v[0];
    for (int i = 1; i < n; i++) if (v[i] > m) m = v[i];
    return m;
}

/** energy = sum(x^2) / n  |  rms = sqrt(energy) */
static float _energy(const float *v, int n)
{
    float s = 0.0f;
    for (int i = 0; i < n; i++) s += v[i] * v[i];
    return s / (float)n;
}

/**
 * Rellena 7 features de un eje en positions [offset .. offset+6]
 * Orden: mean, std, min, max, range, rms, energy
 */
static void _axis_features(const float *v, int n, float *feat, int offset)
{
    float mn  = _mean(v, n);
    float mn2 = _min(v, n);
    float mx  = _max(v, n);
    float en  = _energy(v, n);

    feat[offset + 0] = mn;
    feat[offset + 1] = _std(v, n, mn);
    feat[offset + 2] = mn2;
    feat[offset + 3] = mx;
    feat[offset + 4] = mx - mn2;          /* range */
    feat[offset + 5] = sqrtf(en);         /* rms   */
    feat[offset + 6] = en;                /* energy */
}

/* ── API pública ─────────────────────────────────────────────── */

esp_err_t feature_extractor_extract(
    const float *ax,
    const float *ay,
    const float *az,
    int          n_samples,
    float       *features)
{
    if (n_samples < 2 || ax == NULL || ay == NULL ||
        az == NULL || features == NULL) {
        return ESP_ERR_INVALID_ARG;
    }

    /* ── Features por eje (7 × 3 = 21) ─ */
    _axis_features(ax, n_samples, features,  0);  /* ax → [0..6]  */
    _axis_features(ay, n_samples, features,  7);  /* ay → [7..13] */
    _axis_features(az, n_samples, features, 14);  /* az → [14..20] */

    /* ── Magnitud del vector 3D (3) ─── */
    float mag_sum  = 0.0f;
    float mag_sum2 = 0.0f;
    float mag_max  = 0.0f;

    for (int i = 0; i < n_samples; i++) {
        float m = sqrtf(ax[i]*ax[i] + ay[i]*ay[i] + az[i]*az[i]);
        mag_sum  += m;
        mag_sum2 += m * m;
        if (m > mag_max) mag_max = m;
    }

    float mag_mean = mag_sum / (float)n_samples;
    float mag_var  = (mag_sum2 / (float)n_samples) - (mag_mean * mag_mean);
    float mag_std  = sqrtf(mag_var < 0.0f ? 0.0f : mag_var);

    features[21] = mag_mean;  /* magnitude_mean */
    features[22] = mag_std;   /* magnitude_std  */
    features[23] = mag_max;   /* magnitude_max  */

    /* ── Tasa de cruces por cero de ax (zcr_ax) (1) ─── */
    float ax_mean = features[0];  /* ya calculado */
    int   crossings = 0;
    float prev_centered = ax[0] - ax_mean;

    for (int i = 1; i < n_samples; i++) {
        float curr_centered = ax[i] - ax_mean;
        /* Cruce cuando el signo cambia (excluyendo cero exacto) */
        if ((prev_centered > 0.0f && curr_centered < 0.0f) ||
            (prev_centered < 0.0f && curr_centered > 0.0f)) {
            crossings++;
        }
        prev_centered = curr_centered;
    }

    features[24] = (float)crossings / (float)n_samples;  /* zcr_ax */

    return ESP_OK;
}
